keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+the upstream code.
+
+When opting into using numeric parameters for their DRA driver, driver
+developers no longer need to provide their own controller. Kubernetes takes
+over allocating and deallocating claims.
+
+Scheduling performance is expected to become better compared to using the
+PodSchedulingContext. However, supporting CA is the main reason for this KEP.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Enable cluster autoscaling when pods use resource claims, with correct
+  decisions and changing the cluster size by more than one node at a time.
+
+- Support node-local resources. Adding or removing nodes has no effect
+  on network-attached resources and therefore CA does not need to (and cannot)
+  simulate them.
+
+- Allow DRA driver developers to provide a user experience that is similar to
+  the one possible without numeric parameters. Ideally, users should not notice
+  at all that a driver is using numeric parameters under the hood.
+
+- Support adding and versioning multiple different numeric models without
+  having to modify the implementation of DRA or the numeric parameters
+  extension.
+
+- Keep the core DRA functionality from [KEP
+  #3063](https://github.com/kubernetes/enhancements/issues/3063) working as
+  defined there for Kubernetes 1.29. Cluster autoscaling is not always needed
+  and in-tree numeric model(s) are not expected to cover all use case, so the
+  flexibility offered by opaque parameters remains relevant.
+
+### Non-Goals
+
+- Support numeric class parameters. They remain opaque and can be used to
+  define configuration of resources, but must not affect where a claim can be
+  allocated.
+
+- Define exactly one numeric model that covers all current and future resources.
+  There cannot be a "one size fits all" solution.
+
+## Proposal
+
+### Publishing node capacity
+
+A numeric model defines a capacity type. The type has its own API group, kind
+and potentially different versions. It must support encoding and decoding JSON
+with an apimachinery scheme. The kubelet queries the registered DRA driver
+kubelet plugins for capacity information through a new gRPC call. Drivers
+respond by returning the JSON encoding of this capacity type. A streaming gRPC
+API allows drivers to update the capacity information.
+
+For example, a simple counter could be described as:
+```
+const GroupName = "counter.dra.config.k8s.io"
+const KindName = "Capacity"
+const Version = "v1beta1"
+
+type Capacity struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	metav1.TypeMeta `json:",inline"`
+
+	Count int64 `json:"count"`
+}
```

Comment:
If we need a slice of 1g.20gb then I understand that the `Count` should be 1 but who sets the slice 1g.20gb and how is the node selected?

Reply:
Yes, that's the difference. The "capacity type" is used when publishing capacity of a node, the "parameter type" when asking for resources via the vendor CRD.

Reply:
@tzneal:

> Have you considered expanding metav1.LabelSelector with these features vs. making another label selector?

I have. I'm worried about breaking the existing APIs which use `metav1.LabelSelector`. It's not impossible, but it would have to be done with extreme care. It's probably better to fork, iterate on the extended implementation while it's still alpha, and then later consider merging. Or perhaps not - it would then still be very risky.

@klueska: 

Globbing could be added, it's just another operator.

Could comparison with boolean be handled with comparison against "true"/"false" (strings)? That already works. The vendor CRD might be able to constrain the values for the comparison to just those. If it cannot be expressed by the validation in the CRD, then and admission hook could add this additional validation.

But this brings me to the main issue: arbitrary expressions with brackets. Those are not part of my proposal. If we allow them, how do we represent them? A single string that then gets parsed or nested structures (similar to metav1.LabelSelector)?

https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#set-based-requirement describes a syntax for requirements and https://github.com/kubernetes/apimachinery/blob/02a41040d88da08de6765573ae2b1a51f424e1ca/pkg/labels/selector.go#L212-L267 is similar, but have not found an example of an existing API which accepts such strings. Does anyone know of one?

`LabelSelector` and `FieldSelector` in https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ListOptions might come close, but I think the syntax is different ("fieldA=value1,fieldB=value2").

Reply:
I'm not in favor of arbitrary expressions with brackets. 

This is how I represent them in the current DRA driver for GPUs (where we get to bring along our own API):
https://github.com/NVIDIA/k8s-dra-driver/blob/main/api/utils/selector/selector.go#L31

With a concrete struct that (ultimately) instantiates this type:
https://github.com/NVIDIA/k8s-dra-driver/blob/main/api/nvidia.com/resource/gpu/v1alpha1/gpuselector.go#L32

And an example usage:
https://github.com/NVIDIA/k8s-dra-driver/blob/main/demo/specs/selectors/parameters.yaml#L8

Reply:
> I'm not in favor of arbitrary expressions with brackets.

Agreed. What you have seems usable also for numeric parameters, we just have to generalize.

Reply:
@klueska take a look at this comment if you haven't already: https://github.com/kubernetes/enhancements/pull/4384#issuecomment-1912886646

We shouldn't confuse the example here for the claim parameters with the request objects. I  have an example in there that shows that the "parameter" object, the "request" object and the "capacity" object are different objects, so maybe provides a clearer example. It sounds like based on Patrick's example that includes a selector, you are asking for a selector criteria in the capacity model. That seems a little odd to me, and I would think selector possibly would bubble up to ResourceClaim directly since it's already a well-understood, user-facing concept. Though, I could see a case where either the user or the resource class / driver author would want to specify selectors.

I am curious about the globbing and boolean questions. Booleans are covered in my proposal (as "capabilities"), but right now in there it's just a series of independent capacity constraints AND'd together. One thing I am thinking about for the MIG case is a nested model that provides alternative InstanceCapacity objects that may chosen in a mutually exclusive way. This allows the scheduler and/or cluster autoscaler to iterate through options.

Similarly, you could define a type that provides logical operations (as opposed to just ANDing them all), or we could extend InstanceCapacity with an "expression" field that uses a CEL expression to provide a logical operation on top of the array of ResourceCapacity objects.

Either of those options is *super powerful*. BUT they expand the scheduler / autoscaler search space in likely NP-complete ways. So, we should be careful before going there, and have some mitigation plans - for example, examining the requests and capacity to find the "quick" exclusionary criteria and prune the paths early.

On globbing, I would like to poke at that a little more. It would be easy to support as "just another operator" as described above. But, who is writing those globs? Is it the ResourceClass author or the user consuming the resource? If it's the ResourceClass author, have you considered a controller that applies labels based upon the globs, and then we can rely on ordinary label selection? If we do that, we get the indexing provided by label selection, whereas with globbing we require evaluation against all nodes.

Reply:
I will read through the comment you linked in the morning. I did a review of the KEP text earlier today, but I didn't get to the non-inline comments on the PR yet (which includes your linked comment).

That said, here's a response to your comment here:

> It sounds like based on Patrick's example that includes a selector, you are asking for a selector criteria in the capacity model.

Hmm, I would need to look back at where I left my exact comment, but (as far as my uses cases are concerned) you would *not* need a selector on the `Capacity` object, only the `Parameters` object.

> On globbing, I would like to poke at that a little more. It would be easy to support as "just another operator" as described above. But, who is writing those globs? Is it the ResourceClass author or the user consuming the resource? 

The user consuming the resource in the `ClaimParameters` object he references in his `ResourceClaim`. The capacity object needs to encode the properties of the resources they represent, and the user needs a way to "select" the right resource based on some boolean expression that matches on these properties.

Reply:
See [GPU selection via complex constraints](https://docs.google.com/document/d/1bDO11rEq_Yhpgpk5RN0VwnMLV1_2wNWvtOyM_QoIV_Y/edit#heading=h.io7fw1qvnb8y ) for an example of how I do this in the existing DRA driver for GPUs. We don't have to do things exactly like this, but I want to be able to do something similar.

Reply:
My point about the label selector was that "ClaimParameters" is an example of a vendor-specific CR and so I don't see the point in discussing the details of its contents. It's just a placeholder and not a part of the DRA API. Or am I misundertanding this?

Reply:
If we want selection to work with this numeric models proposal, then there needs to be a generic framework for selection that isn't baked into the vendor specific API.

It's not enough to just say "I have 5 GPUs", you need to be able to say "I have 5 GPUs with this set of properties", and the scheduler needs to be able to match the selector passed in a resource claim to these properties in some way.

Maybe I'm too close to the original DRA proposal and this one doesn't include the notion of a `ClaimParameters` object, but something in the `ResourceClaim` needs to let me pass a selector to match on the properties of the resource type I am requesting.

Reply:
Makes sense. I think it's covered in the proposal, but take a look tomorrow and see what you think. The proposal still has ResourceClaim with a reference to a vendor-specific CR. We translate that into a vendor neutral request object using CEL (which can operate on arbitrary CRs by treating them as `Unstructured`). Various selection criteria could be included there (they are not discussed in there now, beyond a boolean true/false).

Reply:
I agree that we need a more comprehensive expression language than `metav1.LabelSelector`. `metav1.LabelSelector` only gives us `<term1> && <term2> && ...` where each term is one label check.

In @johnbelamaric's proposal, each hardware instance is described by more than one numeric model and the vendor CRD with the claim parameter then gets mapped to one parameter type per model. Each model is then checked independently and all must match an instance.

At first glance this looks appealing because one can "compose" the description of the hardware from different building blocks. But now consider mode switching. Each mode may have labels that need to match the selector provided by the user. The "selector model" is unaware of "modes", so it cannot be used. The "mode switch model" has to re-implement it. Alternatively, more details of the mode switch model need to be hoisted up into the `resource.k8s.io` API. This is a slippery slope where the common API becomes more and more complex.

Therefore I prefer to stick with "one model per driver" and instead implement label selection as a shared library that can be used by the different models. Both the "counter model" and the "mode switch model" would support it, with the difference that the "counter model" publishes a a flat set of labels as part of its capacity while the "mode switch model" has labels for the entire instance and each mode.

Reply:
Regarding globbing -- it's been a long-standing feature request for users to be able to request GPUs based on the model name returned by `nvidia-smi`.

Examples of such model names include the following:
```
Tesla V100-SXM2-16GB
NVIDIA T4 32GB
NVIDIA A100-SXM4-40GB
NVIDIA H100 PCIe
```

As some background, it's common for people to prototype their workloads in `docker` on a specific GPU model before moving them to Kubernetes. When doing so, they are keenly aware of the GPU model they are running on, and want to ensure that they continue to run on this same GPU model once deployed under Kubernetes.

Now keep in mind, this is already possible today using the extended resource exposed by the current device plugin API (i.e. `nvidia.com/gpu`) in conjunction with a node label that advertises the GPU model (assuming all GPUs on a node are homogeneous, of course).

It would look something like this:
```
apiVersion: v1
kind: Pod
metadata:
  name: gpu-example
spec:
  containers:
    - name: gpu-example
      image: nvidia/cuda
      resources:
        limits:
          nvidia.com/gpu: 2
  nodeSelector:
    nvidia.com/gpu.model: NVIDIA A100-SXM4-40GB
```

Independent of the fact that selection occurs at the node level rather than the device level in this example (which DRA proposes to solve), matching on an exact model name is problematic for two reasons:
1. As evidenced above, the format of the model name is inconsistent -- meaning that users have to remember the differences between them when putting their selectors together.
2. There is not just a single model name for, say, an A100 card on a given machine. The model name encodes other information, such as if its on a mother board that connects devices [within a PCIe hierarchy or via SXM](https://gpus.llm-utils.org/nvlink-sxm-and-pcie/#:~:text=SXM%20is%20a%20Nvidia%20specific,SXM%20cards%20are%20NVLink%20compatible), or how much memory the card has on it (e.g. 40GB, 80GB, or 96GB).

Some examples of (2) include:
```
NVIDIA A100-PCIE-40GB
NVIDIA A100-SXM4-40GB
NVIDIA A100-PCIE-64GB
NVIDIA A100-SXM-64GB
NVIDIA A100-PCIE-80GB
NVIDIA A100-SXM4-80GB
``` 

It would make the UX much nicer for one to be able to select on something like "`*A100*` in cases where they only care about getting access to an A100 (and don't care about the interconnect or the amount of memory -- they should be using a separate selector for memory anyway).

For cases where the *do* care about being more specific, they still can be by either matching on the full model name or a different subset of it using an alternate glob expression.

Given the (revised) proposal we have of using a DRA controller to translate any vendor specific selection parameters into selectors that the scheduler will understand (apologies I don't know the right terms here yet to be more specific), I imagine the scheduler could take `modifier` expressions that tell it to ignore case, for example, and the DRA controller could decide if it wants to apply these modifiers or not when performing the translation.

Reply:
> Regarding globbing -- it's been a long-standing feature request for users to be able to request GPUs based on a subset of the model name

People want to glob, substring, prefix, or regex match kube selectors, too.  Just say no.

> The model name encodes other information

You have a stringly-typed structure.  Ideally you would decompose this into actual atoms of information, and let the user select on those. 

> It would make the UX much nicer for one to be able to select on something like "*A100*" in cases where they only care about getting access to an A100

So name the concept that "A100" represents.  "family"?  "generation"?  A glommed together name is great for dumping in a logfile, but we have data structures for a reason.

>  they should be using a separate selector for memory anyway

exactly

Anyway, I probably won't die on the hill of not allowing globbing, but I will push back strongly.